### DISCUSSION TITLE
Unit Test IEventAggregator with Filter?

### DISCUSSION THREAD

**Discussion by wowfood** [PrismLibrary/Prism#2769] [WPF]
Not sure if this is the best place to ask, but bored of getting replies on stack overflow of "Why are you even bothering to test that".  Because my company has an inane idea about 100% coverage, like a load of them do...  

I've found a way to test my event aggregator subscriptions using mstest and moq via the following, I apologize but I need to manually copy the code over so it's poorly formatted and may be missing a bit.

```
MockEventAggregator = new Mock<IEventAggregator>();
MockEvent = new Mock<MyEvent>();
callback = null;

MockEvent.Setup( x=> x.Subscribe( It.IsAny<Action<MyEventArgs>(), It.IsAny<ThreadOptions>(), It.IsAny<bool>(), It.IsAny<Predicate<MyEventArgs>>()))
.Callback<Action<MyEventArgs>, ThreadOption, bool, Predicate<MyEventArgs>((e, t, b, a) => callback = e;
```

I then test the callback with 
`callback.Invoke(new MyEventArgs("Test"));
`
This all works fine, however I'm running into two issues.

First.  If my subscription has a filter on it.  For example x => x.Name == Name then I do not get the code coverage increase I'm expecting, I'm assuming this is because the filter isn't being tested (removing the filter my test coverage increases)

However I am not sure how to write a test with the filter involved.

I've tried setting the predicate line to args => args.Name == "Test"; but then my callback gets set to null.

Just wondering if anyone has had any luck setting up a unit test for the filter portion of the event aggregator subscription.

**Comment by brianlagunas** [comment]
Have you looked at how Prism tests the filter?

https://github.com/PrismLibrary/Prism/blob/master/tests/Prism.Core.Tests/Events/PubSubEventFixture.cs

**Reply by wowfood** [reply]
I've had a quick read through, I think the issue I'm seeing is down to Moq.  I need to mock the events because they're subscribed on the UIThread, otherwise I could use a similar approach with the actual events.  

But the Moq doesn't actually publish the event normally, just triggers the callback event as though the publsih had gone through.

Might just have to explain this one away as not currently testable and we'll have to push the test for this back to when we start automation tests and have the application running.

**Reply by brianlagunas** [reply]
Then don't use Moq for this specific test 🤷 . It's testable.
